purchases/src/main/kotlin/com/revenuecat/purchases/blockstore/BlockstoreHelper.kt

```diff
@@ -0,0 +1,161 @@
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+internal class BlockstoreHelper
+@OptIn(ExperimentalCoroutinesApi::class)
+constructor(
+    applicationContext: Context,
+    private val identityManager: IdentityManager,
+    private val blockstoreClient: BlockstoreClient = Blockstore.getClient(applicationContext),
+    private val ioScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO.limitedParallelism(1)),
+    private val mainScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main),
+) {
+
+    private companion object {
+        const val BLOCKSTORE_USER_ID_KEY = "com.revenuecat.purchases.app_user_id"
+        const val BLOCKSTORE_MAX_ENTRIES = 16
+    }
+
+    fun storeUserIdIfNeeded(customerInfo: CustomerInfo) {
+        val currentUserId = identityManager.currentAppUserID
+        if (
+            !IdentityManager.isUserIDAnonymous(currentUserId) ||
+            customerInfo.allPurchasedProductIds.isEmpty()
+        ) {
+            return
+        }
+        ioScope.launch {
+            val blockstoreData = try {
+                getBlockstoreData()
+            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
+                errorLog(e) { "Failed to retrieve Block store data. Will not store userId. Error: ${e.message}" }
+                return@launch
+            }
+            try {
+                storeUserIdIfNeeded(blockstoreData, currentUserId)
+            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
+                errorLog(e) { "Failed to store user Id in Block store: ${e.message}" }
+            }
+        }
+    }
+
+    fun aliasCurrentAndStoredUserIdsIfNeeded(callback: () -> Unit) {
+        fun callCompletion() {
+            mainScope.launch {
+                callback()
+            }
+        }
+        val currentUserId = identityManager.currentAppUserID
+        if (!IdentityManager.isUserIDAnonymous(currentUserId)) {
+            callCompletion()
+            return
+        }
+        ioScope.launch {
+            val blockstoreData = try {
+                getBlockstoreData()
+            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
+                errorLog(e) { "Failed to retrieve Block store data. Will not recover userId. Error: ${e.message}" }
+                callCompletion()
+                return@launch
+            }
+            val blockstoreUserId = blockstoreData[BLOCKSTORE_USER_ID_KEY]?.bytes?.let { String(it) }
+            if (blockstoreUserId == null || blockstoreUserId == currentUserId) {
+                callCompletion()
+                return@launch
+            }
+            try {
+                debugLog { "Aliasing Blockstore user ID: $blockstoreUserId with current UserID" }
+                identityManager.aliasCurrentUserIdTo(
```

Comment:
this made me think that this flow will only work for devs with the default or legacy transfer behavior https://www.revenuecat.com/docs/projects/restore-behavior#transfer-to-new-app-user-id. **Transfer if there are no active subscriptions** won't work because the non-consumable would be active in the previous old customer

Reply:
ahh right, `/alias` doesn't receive any purchases so it just aliases users. You're correct. I had assumed it behave the same as `/receipts` but it won't since it doesn't get any purchase